OpenAPI specs for API

ilammy opened the pull request:
This is a preview version of OpenAPI (aka Swagger) specification of node API extracted from current documentation and source code. This machine-readable format is going to replace the current informal specification of the API in documentation.

It covers the following services:

* Exonum node management
* Exonum blockchain explorer
* BTC anchoring service
* configuration update service
* time oracle service

The content is rather crude, the descriptions are not fully complete, there are some spelling/grammar issues, some examples are missing. But it should be technically accurate for Exonum 0.9.0 (as in describing all endpoints, parameters, and results).

The API is currently not rendered anywhere. You can get a preview with [online OpenAPI editor](https://editor.swagger.io), just paste raw YAML content there. Please refer to those previews and provide feedback on what could be improved from technical standpoint.

MartaDubno commented:
Looks good. I found two technical issues in the UI part of the editor, not sure if its the problem with the code or the editor. The issues are shown on the screenshot below.
- This relates to all the files and the Models sections, specifically the first 2 or 4 models in them. Once I open them I cannot close these models, the button isn't there.
- In a non-full screen mode, long descriptions and examples go beyond the grey fields where they belong.
![screen shot 2018-07-31 at 5 07 07 pm 2](https://user-images.githubusercontent.com/36300666/43466841-1fbe4858-94e9-11e8-956a-4c5a59c5e4c6.png)
.

ilammy commented:
API spec will need an update once exonum/exonum#839 is merged.

slowli commented on src/advanced/blockchain-explorer.yaml:
```diff
@@ -0,0 +1,528 @@
+                    type: array
+                    description: The list of the transactions included into the block.
+                    items:
+                      $ref: "#/components/schemas/SerializedTransaction"
```
I think the transaction hashes are used here.

ilammy replied:
Yeah, my bad. Thanks for having hawk eyes.

This object is actually serialized from this struct:

```rust
pub struct BlockInfo {
    pub block: Block,
    pub precommits: Vec<Precommit>,
    pub txs: Vec<Hash>,
}
```

slowli commented on src/advanced/blockchain-explorer.yaml:
```diff
@@ -0,0 +1,528 @@
+      - committed
+      - in-pool
+      - unknown
+    TransactionStatus:
```
Maybe, it could make more sense to represent the status as `oneOf` with a discriminator?

ilammy replied:
I'm not sure. It does make sense from the rigorous point of view, but I kinda doubt that we'll get any merit from it.

`oneOf` does not seem to allow to specify which discriminator values choose which variants. It is a way to join two very different schemas into one. It is used for `KnownTransaction` here (and could have been used for `UnknownTransaction` as well), but that's because committed and uncommitted transactions have very different representations.

In this case we still have a single `TransactionStatus` type with a single purpose. The most common way of handling it should be check the `type` field for `"success"` and if it's not it then propagate information from `code` and `description`. I think we can leave it as a single type with some meaningful default values for `code` and `description` and don't bother with accurate discriminated union representation.

Though, here's how it's represented in the code:

```rust
enum TxStatus<'a> {
    Success,
    Panic { description: &'a str },
    Error { code: u8, description: &'a str },
}
```

so I guess we should translate that directly into `oneOf`. Maybe some day we'll have OpenAPI docs generated from the code. If I were the generator I would translate that `enum` as `oneOf`.

slowli commented on src/advanced/node-management.yaml:
```diff
@@ -0,0 +1,361 @@
+
+components:
+  schemas:
+    NullObject:
```
Looks clunky, but I'm not sure if a better type description is possible (e.g., `type: "null"` doesn't seem to work).

ilammy replied:
Yeah. For some reason OpenAPI does not have `type: null` and this seems to be the closest we can get to 'only `null` is allowed here'. Though, technically this type specification allows any object in addition to `null`.

slowli commented on src/advanced/time.yaml:
```diff
@@ -0,0 +1,120 @@
+      tags: [Private API]
+      summary: Get timestamps of all validators
+      description: >
+        Returns the latest timestamps indicated by all validator nodes for which time is known.
```
Nit: I'd elaborate that by all validator nodes we mean both current and past ones.

ilammy commented:
Added a bunch of changes to resolve issues pointed out by @slowli.

One thing that bothers me is the `maximum` values for some types. We use `u64` in Rust code so I have added appropriate limits:

```yaml
minimum: 0
maximum: 18446744073709551615  # <== 2^64 - 1
```

However, when rendered by Swagger the limit it shown as 18446744073709552000. I guess Swagger can't handle unsigned 64-bit values accurately. It's not possible to specify the limit as a string, it has to be a number.

ilammy commented:
@MartaDubno Uncollapsible model entries seem to be a bug in Swagger's editor.

As for the examples... Do we need to keep accurate examples in data model sections? They do not seem to word-wrap correctly. Though JSON objects in request examples wrap normally. I guess we can remove the model value examples and keep them only in requests.

ilammy commented:
Can we merge this please?

slowli commented on src/advanced/bitcoin-anchoring.yaml:
```diff
@@ -0,0 +1,259 @@
+          schema:
+            type: integer
+            minimum: 0
+            maximum: 4294967295
```
[Validator is represented by `u16`](https://docs.rs/exonum/0.8.0/exonum/helpers/struct.ValidatorId.html), so the upper limit should be 65535.

slowli commented on src/advanced/bitcoin-anchoring.yaml:
```diff
@@ -0,0 +1,259 @@
+              schema:
+                type: string
+                nullable: true
+                format: "^[0-9a-f]+$"
```
Are you sure this works? I was under the impression (reading the OpenAPI spec) that [ECMA-262 regexps](https://www.ecma-international.org/ecma-262/5.1/#sec-7.8.5) should be used (i.e., ones used in JavaScript; something like `/^[0-9a-f]+$/i`). Additionally, shouldn't the property be `pattern`, rather than `format`?

slowli commented on src/advanced/blockchain-explorer.yaml:
```diff
@@ -0,0 +1,558 @@
+        tx_count:
+          type: integer
+          minimum: 0
+          maximum: 4294967295
```
I don't think it makes sense to include `u32` or `u64` upper restrictions, tbh, especially for returned types. Additionally, `u64::max_value()` isn't even represented properly in the Swagger doc viewer.
